### DISCUSSION TITLE
Web UI port and proxy server port

### DISCUSSION THREAD

**Discussion by Zhiw** [gravitational/teleport#30539] [Q&A] [answered]
I have a lot of servers, I want to connect to them via Teleport SSH, but I'll set up security groups for security purposes, and I want my Web UI to be accessible only by certain IPs, but it seems that the proxy server and the Web UI use the same port, but my servers are dynamically being added, so is there anything I can do to solve this problem?

**Comment by webvictim** [comment, marked as answer]
Unfortunately IP whitelisting and dynamic IP addressing aren't really compatible. You'll need to choose one or the other. Teleport's web UI and API ports are the same.

You could put a WAF or ALB in front of the Teleport proxy if you wanted which could enforce things like country-based restrictions or connection limits.

For what it's worth, we consider the Teleport proxy service to be secure enough to be deployed open to the internet with no incoming IP restrictions. This is how our SaaS is successfully deployed for all customers. Teleport does have some built-in rate limiting of its own.

**Reply by Zhiw** [reply]
Thank you🤗

**Comment by Zhiw** [comment]
Thanks for your suggestion, I want to configure an ALB now, how should I configure the ALB and teleport, like listening configuration, forwarding rules. 
And I got this logs:
 'Closing HTTP connection: HTTP listener is disabled'
'Detected an HTTP request. If this is for a health check, use an HTTPS request instead'
Thanks.

**Reply by webvictim** [reply]
You should make sure that the target group your ALB is forwarding to is using the HTTPS protocol.

![image](https://github.com/gravitational/teleport/assets/897821/2b322fc0-5e30-4776-ac95-8cf1b83c7958)

**Reply by Zhiw** [reply]
HTTPS(Load Balancer) -> Teleport(HTTPS)?
Based on your advice, I've successfully deployed it. 
One thing I need to confirm is that I do not need to put my certificate on the teleport instance, teleport will issue it itself, right? 
I can now SSH to my node from the web, but when I use `tsh login`, I got an error: `x509: certificate signed by unknown authority`. I see that this issue is also being discussed 
 here: https://github.com/gravitational/teleport/discussions/13522, may I ask whether this issue has been found and solved.

**Reply by webvictim** [reply]
Correct, you don't need to provide a certificate to Teleport - the load balancer will handle TLS for you (with a certificate from ACM or similar)

What versions of Teleport and tsh are you using?

`teleport version`
`tsh version`

You'll need both to be version 13 for Teleport to work behind an ALB.

**Reply by Zhiw** [reply]
If I put the certificate on both the load balancer and the teleport, it is also feasible？Or maybe it's not really necessary.
Sorry, I upgrade the version of tsh, it worked.

**Reply by webvictim** [reply]
You can use the same certificate for Teleport too if you have a way of doing that!

**Reply by Zhiw** [reply]
I have two teleport instances behind ALB, when I SSH to my node, it says 'no tunnel connection found', but when I connect again, it connects.
I think it's because the ssh node has a reverse tunnel with node 1, but the first time I SSH, it connects to node 2, so I can not SSH to it. And then the second time, the traffic goes to the node 1, so it's connected.
What should I do to fix this?

**Reply by webvictim** [reply]
Try enabling "cross-zone load balancing" in your ALB settings. At the moment your node is only able to create a tunnel to one of your two proxies.

If you're using Teleport Enterprise, the other option is to set up [proxy peering](https://goteleport.com/docs/architecture/proxy-peering) in your cluster.

**Comment by Zhiw** [comment]
Do you mean that after enabling the settings, the node will create tunnels to both proxies at the same time?
Or is it only possible to create a tunnel by design？

**Reply by webvictim** [reply]
Yes, the former. Teleport's discovery mechanism works by querying for the list of all proxies joined to the cluster by UUID, then repeatedly establishing connections to the proxy load balancer to establish tunnels until it has "seen" all the proxy UUIDs.

In situations where cross-zone load balancing is not enabled, Teleport will only ever be able to create a tunnel to one of the two proxies because the load balancer will never send traffic to the second proxy. This seems to be what's happening in your situation, hence why connections only work 50% of the time.

**Comment by kmmanto** [comment]
@webvictim Is this supported in GCP GKE?

Application LB (with ManagedCertificate) -> teleport 

1) HTTPS Healthcheck fails with 

```
2023-08-24T16:29:50Z DEBU [AUTH]      ClientCertPool -> cert(teleport.example.com issued by teleport.example.com:164478893243978877637117510850908878079) auth/middleware.go:760
2023-08-24T16:29:50Z DEBU [AUTH]      ClientCertPool -> cert(teleport.example.com issued by teleport.example.com:14057231848138938509319863491798109801) auth/middleware.go:760
2023-08-24T16:29:50Z WARN [ALPN:PROX] Failed to handle client connection. error:[
ERROR REPORT:
Original Error: *errors.errorString acme/autocert: missing server name
Stack Trace:
        github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:392 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).handleConn
        github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:326 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).Serve.func1
        runtime/asm_amd64.s:1598 runtime.goexit
User Message: acme/autocert: missing server name] alpnproxy/proxy.go:337
2023-08-24T16:29:51Z DEBU [AUTH]      ClientCertPool -> cert(teleport.example.com issued by teleport.example.com:164478893243978877637117510850908878079) auth/middleware.go:760
2023-08-24T16:29:51Z DEBU [AUTH]      ClientCertPool -> cert(teleport.example.com issued by teleport.example.com:14057231848138938509319863491798109801) auth/middleware.go:760
2023-08-24T16:29:51Z WARN [ALPN:PROX] Failed to handle client connection. error:[
ERROR REPORT:
Original Error: *errors.errorString acme/autocert: missing server name
Stack Trace:
        github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:392 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).handleConn
        github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:326 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).Serve.func1
        runtime/asm_amd64.s:1598 runtime.goexit
User Message: acme/autocert: missing server name] alpnproxy/proxy.go:337
```
Note that teleport is not running with any certifcates.

2) HTTP Healtcheck
Same error as above.
```
2023-08-24T16:29:22Z DEBU [MX:PROXY:] Closing HTTP connection: HTTP listener is disabled. multiplexer/multiplexer.go:287
2023-08-24T16:29:28Z DEBU [MX:PROXY:] Detected an HTTP request. If this is for a health check, use an HTTPS request instead. multiplexer/multiplexer.go:285
2023-08-24T16:29:28Z DEBU [MX:PROXY:] Closing HTTP connection: HTTP listener is disabled. multiplexer/multiplexer.go:287
```

**Reply by webvictim** [reply]
If you're terminating TLS at the load balancer level and Teleport doesn't need a certificate, set `acme.enabled` to `false` in `/etc/teleport.yaml` and restart.

**Comment by Zhiw** [comment]
I deployed two teleports and then mounted one NAS, and they shared a data directory, so the host_uuid was the same. So for nodes, there is only one proxy in the cluster with same UUID.
There are no suitable backend options in our scenario, so we chose to mount the same NAS.
Do you have any good suggestions？
Thank you very much for your answer.
